docs/ref-manual/cvl/methods.md

```diff
@@ -46,44 +55,280 @@ method_summary   ::= "ALWAYS" "(" value ")"
                    | "DISPATCHER" [ "(" ( "true" | "false" ) ")" ]
                    | "AUTO"
-                   | [ "with" "(" cvl_params ")" ] block
-                   | [ "with" "(" cvl_params ")" ] expression
+                   | [ "with" "(" "env" id ")" ] block
+                   | [ "with" "(" "env" id ")" ] expression
+                   | [ "with" "(" "env" id ")" ] id "(" [ id { "," id } ] ")"
 
-cvl_param ::= cvl_type [ id ]
+```
 
+```{todo}
+The `with(env id) block` and `with(env id) expression` productions come from
+the grammar, but I suspect the only valid block/expression summary is the third
+option (function/ghost summaries).  See last section of this document.
 ```
 
 See {doc}`types` for the `evm_type` and `cvl_type` productions.  See {doc}`basics`
 for the `id` production.  See {doc}`statements` for the `block` production, and
 {doc}`expr` for the `expression` production.
 
+(envfree)=
+Entries in the `methods` block
+------------------------------
+
+Each entry in the methods block denotes either the sighash or the ABI signature
+for a contract method.  Methods of contracts that are introduced by {doc}`using
+statements <using>` can also be described by prefixing the method name with
+the contract variable name.  For example, if contract `C` is introduced by the
+statement `using C as c`, then the method `f(uint)` of contract `c` can be
+referred to as `c.f(uint)`.
+
+It is possible for a method signature to appear in the `methods` block but not
+in the contract being verified.  In this case, the prover will skip any rules
+that mention the missing method, rather than reporting an error.  This behavior
+allows reusing specifications on contracts that only support part of an
+interface: only the supported methods will be verified.
+
+Following the method signature is an optional `returns` clause.  If a method
+declaration contains a `returns` clause, the declared return type must match
+the contract method's return type.  If the `returns` clause is omitted, the
+return type is taken from the contract method's return type.
+
+Following the `returns` clause is an optional `envfree` tag.  Marking a method
+with `envfree` has two effects.  First, {ref}`calls <call-expr>` to the method
+from CVL do not need to explicitly pass an {term}`environment` value as the
+first argument.  Second, the prover will verify that the method implementation
+in the contract being verified does not depend on any of the environment
+variables.  The results of this check are displayed on the verification report
+as a separate rule called `envfreeFuncsStaticCheck`.
+
 ```{todo}
-This document is incomplete.  See {doc}`/docs/confluence/advanced/methods-overview`,
-{doc}`/docs/confluence/advanced/summaries`, {doc}`/docs/confluence/advanced/internal-summaries`,
-and {doc}`/docs/confluence/advanced/expressive-summaries` for the old documentation
-about the methods block.
+There is a separate check called `envfreeFuncsAreNonpayable`.  Why is this
```

Comment:
I always assumed the "static check" didn't require a call to z3 while the other envfree rules did?

Reply:
The question is why do we need a check for nonpayability?

Reply:
I understand now - payable functions depend on the msg.value which is part of the environment.  Updated.